File: qasdk/src/main/java/co/astrnt/qasdk/core/AstronautApi.java
package co.astrnt.qasdk.core;

import android.content.res.Resources;
import android.os.Build;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.itkacher.okhttpprofiler.OkHttpProfilerInterceptor;

import java.util.concurrent.TimeUnit;

import co.astrnt.qasdk.ApiService;
import co.astrnt.qasdk.dao.CustomFieldApiDao;
import co.astrnt.qasdk.dao.CustomFieldDeserializer;
import co.astrnt.qasdk.dao.InformationApiDao;
import co.astrnt.qasdk.dao.InformationDeserializer;
import co.astrnt.qasdk.dao.SummaryQuestionApiDao;
import co.astrnt.qasdk.dao.SummaryQuestionDeserializer;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by deni rohimat on 06/04/18.
 */
public class AstronautApi {

    private ApiService mApiService;

    public AstronautApi(String baseUrl, boolean isDebugable) {

        final String manufacturer = Build.MANUFACTURER;
        final String model = Build.MODEL;
        final String device = String.format("%s %s", manufacturer, model);
        final String os = "Android " + Build.VERSION.RELEASE;

        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
        httpClientBuilder.readTimeout(60, TimeUnit.SECONDS);
        httpClientBuilder.connectTimeout(60, TimeUnit.SECONDS);
        httpClientBuilder.addInterceptor(chain -> {
            Request request = chain.request().newBuilder()
                    .addHeader("device", device)
                    .addHeader("os", os)
                    .addHeader("browser", "")
                    .addHeader("screenresolution", getScreenWidth() + "x" + getScreenHeight())
                    .build();
            return chain.proceed(request);
        });

        if (isDebugable) {
            HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
            loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

            httpClientBuilder.addInterceptor(loggingInterceptor);
            httpClientBuilder.addInterceptor(new OkHttpProfilerInterceptor());
        }

        Gson gson = new GsonBuilder()
                .registerTypeAdapter(InformationApiDao.class, new InformationDeserializer())
                .registerTypeAdapter(CustomFieldApiDao.class, new CustomFieldDeserializer())
                .registerTypeAdapter(SummaryQuestionApiDao.class, new SummaryQuestionDeserializer())
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(httpClientBuilder.build())
                .baseUrl(baseUrl)
                .build();

        mApiService = retrofit.create(ApiService.class);
    }

    private static int getScreenWidth() {
        return Resources.getSystem().getDisplayMetrics().widthPixels;
    }

    private static int getScreenHeight() {
        return Resources.getSystem().getDisplayMetrics().heightPixels;
    }

    public ApiService getApiService() {
        return mApiService;
    }

}

File: qasdk/src/main/java/co/astrnt/qasdk/core/AstronautApi.kt
package co.astrnt.qasdk.core

import android.content.Context
import android.content.res.Resources
import android.os.Build
import co.astrnt.qasdk.ApiService
import co.astrnt.qasdk.dao.*
import com.google.gson.GsonBuilder
import com.localebro.okhttpprofiler.OkHttpProfilerInterceptor
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.net.URISyntaxException
import java.util.concurrent.TimeUnit


class AstronautApi(baseUrl: String, isDebugable: Boolean, context: Context) {
    val apiService: ApiService

    companion object {
        private val screenWidth: Int
            get() = Resources.getSystem().displayMetrics.widthPixels
        private var newBaseUrl: String = ""
        private val screenHeight: Int
            get() = Resources.getSystem().displayMetrics.heightPixels
    }

    init {
        val manufacturer = Build.MANUFACTURER
        val model = Build.MODEL
        val device = String.format("%s %s", manufacturer, model)
        val os = "Android " + Build.VERSION.RELEASE
        newBaseUrl = baseUrl
        val httpClientBuilder = OkHttpClient.Builder()
        httpClientBuilder.readTimeout(60, TimeUnit.SECONDS)
        httpClientBuilder.connectTimeout(60, TimeUnit.SECONDS)
        httpClientBuilder.addInterceptor(Interceptor { chain: Interceptor.Chain ->
            val request = chain.request().newBuilder()
                .addHeader("device", device)
                .addHeader("os", os)
                .addHeader("browser", "")
                .addHeader("screenresolution", screenWidth.toString() + "x" + screenHeight)
                .build()
            chain.proceed(request)
        })

        httpClientBuilder.addInterceptor(Interceptor { chain: Interceptor.Chain ->
            var requestURL = chain.request()
            var newUrl: HttpUrl? = null
            if (newBaseUrl.toHttpUrlOrNull() != null) {
                try {
                    newUrl = requestURL.url.newBuilder()
                        .scheme(newBaseUrl.toHttpUrlOrNull()!!.scheme)
                        .host(newBaseUrl.toHttpUrlOrNull()!!.toUrl().toURI().host)
                        .build()
                } catch (e: URISyntaxException) {
                    e.printStackTrace()
                }
            }
            assert(newUrl != null)
            requestURL = requestURL.newBuilder()
                .url(newUrl.toString())
                .build()
            chain.proceed(requestURL)
        })
        if (isDebugable) {
            httpClientBuilder.addInterceptor(OkHttpProfilerInterceptor())
        }
//        httpClientBuilder.addInterceptor(ChuckerInterceptor.Builder(context).build())
        val gson = GsonBuilder()
            .registerTypeAdapter(InformationApiDao::class.java, InformationDeserializer())
            .registerTypeAdapter(CustomFieldApiDao::class.java, CustomFieldDeserializer())
            .registerTypeAdapter(
                SummaryQuestionApiDao::class.java,
                SummaryQuestionDeserializer()
            )
            .create()
        val retrofit = Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create(gson))
            .client(httpClientBuilder.build())
            .baseUrl(newBaseUrl)
            .build()
        apiService = retrofit.create(ApiService::class.java)
    }
}